src/aspire/abinitio/commonline_sync3n.py

```diff
@@ -0,0 +1,1025 @@
+
+        :return: An array of length n-choose-2 consisting of 1 or -1, where the sign of the
+        i'th entry indicates whether the i'th relative orientation matrix will be J-conjugated.
+        """
+
+        logger.info(
+            "Initiating power method to estimate J-synchronization matrix eigenvector."
+        )
+        # Set power method tolerance and maximum iterations.
+        epsilon = self.epsilon
+        max_iters = self.max_iters
+
+        # Initialize candidate eigenvectors
+        n_Rijs = Rijs.shape[0]
+        vec = randn(n_Rijs, seed=self.seed)
+        vec = vec / norm(vec)
+        residual = 1
+        itr = 0
+
+        # Todo
+        # I don't like that epsilon>1 (residual) returns signs of random vector
+        #      maybe force to run once? or return vec as zeros in that case?
+        #      Seems unintended, but easy to do.
+
+        # Power method iterations
+        while itr < max_iters and residual > epsilon:
+            itr += 1
+            # Todo, this code code actually needs double precision for accuracy... forcing.
+            vec_new = self._signs_times_v(Rijs, vec).astype(np.float64, copy=False)
+            vec_new = vec_new / norm(vec_new)
+            residual = norm(vec_new - vec)
+            vec = vec_new
+            logger.info(
+                f"Iteration {itr}, residual {round(residual, 5)} (target {epsilon})"
+            )
+
+        # We need only the signs of the eigenvector
+        J_sync = np.sign(vec)
+
+        return J_sync
+
+    def _signs_times_v(self, Rijs, vec):
+        """
+        Multiplication of the J-synchronization matrix by a candidate eigenvector `vec`
+
+        Wrapper for cpu/gpu dispatch.
+
+        :param Rijs: An n-choose-2x3x3 array of estimates of relative rotations
+        :param vec: The current candidate eigenvector of length n-choose-2 from the power method.
+        :return: New candidate eigenvector.
+        """
+        # host/gpu dispatch
+        if self._gpu_module:
+            new_vec = self._signs_times_v_cupy(Rijs, vec)
+        else:
+            new_vec = self._signs_times_v_host(Rijs, vec)
+
+        return new_vec.astype(vec.dtype, copy=False)
+
+    def _signs_times_v_host(self, Rijs, vec):
+        """
+        See `_signs_times_v`.
+
+        CPU implementation.
+        """
+
+        new_vec = np.zeros_like(vec)
+
+        _signs_confs = np.array(
+            [[1, 1, 1], [-1, 1, -1], [-1, -1, 1], [1, -1, -1]], dtype=int
+        )
+
+        c = np.empty((4))
+        desc = "Computing signs_times_v"
+        if self.J_weighting:
+            desc += " with J_weighting"
+        for i in trange(self.n_img, desc=desc):
+            for j in range(
+                i + 1, self.n_img - 1
+            ):  # check bound (taken from MATLAB mex)
+                ij = self._pairs_to_linear[i, j]
+                Rij = Rijs[ij]
+                for k in range(j + 1, self.n_img):
```

Comment:
Same remark as above for all triplet loop

Reply:
`-2`'d it